haproxy_stats.sh

```diff
@@ -18,31 +18,71 @@ pxname="$1"
 svname="$2"
 stat="$3"
 
-DEBUG=${DEBUG:-0}
-HAPROXY_SOCKET="${HAPROXY_SOCKET:-/var/run/haproxy/info.sock}"
+SCRIPT_DIR=`dirname $0`
+CONF_FILE="${SCRIPT_DIR}/haproxy_zbx.conf"
+CONF_MAP="
+ha_sock=/var/run/haproxy/info.sock
+ha_stats_cache_file=/var/tmp/haproxy_stat.cache
+ha_stats_cache_expire=5
+ha_info_cache_file=/var/tmp/haproxy_info.cache
+ha_info_cache_expire=5
+ha_info_cache_expires=6
+debug=0
+debug_only_log=0
+stats_log_file=/var/tmp/haproxy_stat.log
+"
+get_conf() {
+    local conf_val=`grep $1= ${CONF_FILE} | grep -v "^#" | xargs -d "=" | grep $1 |awk '{print $2}'`
+    if [ $conf_val ]; then
+        echo $conf_val
+    else
+        echo "`echo -e "$CONF_MAP" | grep $1=  | xargs -d "=" | grep $1 |awk '{print $2}'`"
+    fi
+}
+
+
+DEBUG=${DEBUG:-`get_conf debug`}
+DEBUG_ONLY_LOG=${DEBUG_ONLY_LOG:-`get_conf debug_only_log`} # only debug in logfile
+HAPROXY_SOCKET="${HAPROXY_SOCKET:-`get_conf ha_sock`}"
 QUERYING_METHOD="${QUERYING_METHOD:-SOCKET}"
-CACHE_STATS_FILEPATH="${CACHE_STATS_FILEPATH:-/var/tmp/haproxy_stats.cache}"
-CACHE_STATS_EXPIRATION="${CACHE_STATS_EXPIRATION:-5}" # in minutes
-CACHE_INFO_FILEPATH="${CACHE_INFO_FILEPATH:-/var/tmp/haproxy_info.cache}" ## unused
-CACHE_INFO_EXPIRATION="${CACHE_INFO_EXPIRATION:-5}" # in minutes ## unused
+CACHE_STATS_FILEPATH="${CACHE_STATS_FILEPATH:-`get_conf ha_stats_cache_file`}"
+CACHE_STATS_EXPIRATION="${CACHE_STATS_EXPIRATION:-`get_conf ha_stats_cache_expire`}" # in minutes
+CACHE_INFO_FILEPATH="${CACHE_INFO_FILEPATH:-`get_conf ha_info_cache_expire`}" ## unused
+CACHE_INFO_EXPIRATION="${CACHE_INFO_EXPIRATION:-`get_conf ha_info_cache_expire`}" # in minutes ## unused
+LOG_FILE="`get_conf stats_log_file`"
 GET_STATS=${GET_STATS:-1} # when you update stats cache outsise of the script
 SOCAT_BIN="$(which socat)"
 NC_BIN="$(which nc)"
-FLOCK_BIN="$(which flock)"
-FLOCK_WAIT=15 # maximum number of seconds that "flock" waits for acquiring a lock
-FLOCK_SUFFIX='.lock'
-CUR_TIMESTAMP="$(date '+%s')"
+
+LOG_TMP=""
+[ "${DEBUG}" -eq 1 ] && LOG_TMP=${LOG_TMP}"\n######__________ $(date) ____________\n" || true
 
 debug() {
-  [ "${DEBUG}" -eq 1 ] && echo "DEBUG: $@" >&2 || true
+  if [ "${DEBUG}" -eq 1 ]; then
+      if [ "${DEBUG_ONLY_LOG}" -ne 1 ]; then
+          echo "DEBUG: $@"
+      fi
+      LOG_TMP=${LOG_TMP}"\nDEBUG: $@"
+  fi
+}
+
+write_log() {
+    echo -e ${LOG_TMP} >> ${LOG_FILE}
+}
+
+fail() {
+    echo $2
+    debug $2
+    write_log
+    exit $1
 }
```

Comment:
the `write_log` logic is a bit too complicated.. let's try to simplify it
```
debug() {
  [[ "${DEBUG}" -eq 1 ]] || true  # return immediately if debug is disabled
  if [[ "${DEBUG_ONLY_LOG}" -eq 1 ]]; then
     "DEBUG: $@" >> ${LOG_FILE}
     return
   else
     echo >&2 "DEBUG: $@"
   fi
}

fail() {
  local _exit_code=${1:-1}
  shift 1
  if [[ -n "$1" ]]; then
    if [[ "${DEBUG}" -eq 0 ]]; then
      echo >&2 "$@"
    else
      debug "$@"
    fi
  fi
  exit $_exit_code
}
```
would this do what you are looking for?

Reply:
if you'd like to prepend your debug log file with a date, it can be added separately when script starts
```
debug "######__________ $(date) ____________"
```

Reply:
I like the removing of the `write_log` and the new `fail` implementation but regarding the `debug` I think the following should do a better job:
```
debug() {
    [[ "${DEBUG}" -eq 1 ]] || return  # return immediately if debug is disabled
    echo "DEBUG: $@" >> ${LOG_FILE}
    [[ ${DEBUG_ONLY_LOG} -ne 1 ]] || return  # return if it should only debug to log
    echo >&2 "DEBUG: $@"
}
```

Reply:
agreed, it's cleaner that way, though I think you've meant to have `-eq` when checking `$DEBUG_ONLY_LOG` value.. as if DEBUG_ONLY_LOG is enabled, there is no reason to write to stderr
```
debug() {
    [[ "${DEBUG}" -eq 1 ]] || return  # return immediately if debug is disabled
    echo "DEBUG: $@" >> ${LOG_FILE}
    [[ ${DEBUG_ONLY_LOG} -eq 1 ]] || return  # return if it should only debug to log
    echo >&2 "DEBUG: $@"
}
```

Reply:
misread my own logic.. nevermind )